### PR TITLE
[Merged by Bors] - feat: Define the covolume of a `ZLattice`

### PR DESCRIPTION
**Pull request by xroblot**
This PR defines the covolume of a `Zlattice` as the (finite) volume of any of its fundamental domain and proves some results about it, mainly: it's nonzero, it does not depend of the choice of the fundamental domain and it is equal to the absolute value of the determinant of any basis of the lattice. 

This PR also creates a directory `Zlattice` (with the original file becoming `Zlattice/Basic.lean`) and a new file `Zlattice/Covolume.lean`.

---

- [x] depends on: #11323 
- [x] depends on: #11321 

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/MeasureTheory/Measure/Haar/OfBasis.lean [hunk 1]
```diff
@@ -54,6 +54,16 @@ theorem mem_parallelepiped_iff (v : ι → E) (x : E) :
...
 #align mem_parallelepiped_iff mem_parallelepiped_iff
 
+theorem parallelepiped_basis_eq (b : Basis ι ℝ E) :
+    parallelepiped b = {x | ∀ i, b.repr x i ∈ Set.Icc 0 1} := by
+  classical
+  ext x
+  simp_rw [mem_parallelepiped_iff, mem_setOf_eq, b.ext_elem_iff, _root_.map_sum,
+    SMulHomClass.map_smul, Finset.sum_apply', Basis.repr_self, Finsupp.smul_single, smul_eq_mul,
+    mul_one, Finsupp.single_apply, Finset.sum_ite_eq', Finset.mem_univ, ite_true, mem_Icc,
+    Pi.le_def, Pi.zero_apply, Pi.one_apply, ← forall_and]
```
Comment (faenuccio) [inline Mathlib/MeasureTheory/Measure/Haar/OfBasis.lean:64]:
```suggestion
  rfl
```

**Review by riccardobrasca — APPROVED** [review]
This is surely my fault as I don't know the measure theory part of the library, but can you clarify the role of `μ`?

**Comment by xroblot** [comment]
> This is surely my fault as I don't know the measure theory part of the library, but can you clarify the role of `μ`?

I am not sure what you mean. The covolume is defined as the volume of any fundamental domain so it is necessary to provide a measure since the hypothesis we have on the ambient space `E` does not ensure that there is a canonical measure on it. It is true that, in most practical cases, `E` will be a `MeasureSpace` so we just want to use the volume of `E`. The tactic `volume_tac` does that automatically is `μ` is not provided.

**Comment by riccardobrasca** [comment]
> > This is surely my fault as I don't know the measure theory part of the library, but can you clarify the role of `μ`?
> 
> I am not sure what you mean. The covolume is defined as the volume of any fundamental domain so it is necessary to provide a measure since the hypothesis we have on the ambient space `E` does not ensure that there is a canonical measure on it. It is true that, in most practical cases, `E` will be a `MeasureSpace` so we just want to use the volume of `E`. The tactic `volume_tac` does that automatically is `μ` is not provided.

So in practice if I have a lattice in, say, `ℂ` and I want to use the standard Lebesgue measure I have to write it explicitly?

**Comment by xroblot** [comment]
> > > This is surely my fault as I don't know the measure theory part of the library, but can you clarify the role of `μ`?
> > 
> > 
> > I am not sure what you mean. The covolume is defined as the volume of any fundamental domain so it is necessary to provide a measure since the hypothesis we have on the ambient space `E` does not ensure that there is a canonical measure on it. It is true that, in most practical cases, `E` will be a `MeasureSpace` so we just want to use the volume of `E`. The tactic `volume_tac` does that automatically is `μ` is not provided.
> 
> So in practice if I have a lattice in, say, `ℂ` and I want to use the standard Lebesgue measure I have to write it explicitly?

Well, we could write `covolume L volume` (in fact, `MeasureTheory.MeasureSpace.volume`) or just `covolume L` (as it is written in `covolume_eq_det`) since Lean can infer automatically what the measure is since it is a `MeasureSpace`. It is only if you want to use an "unusual" measure (or the space `E` has no "canonical" measure) that would actually need to explicit provide it.

**Comment by riccardobrasca** [comment]
> > > > This is surely my fault as I don't know the measure theory part of the library, but can you clarify the role of `μ`?
> > > 
> > > 
> > > I am not sure what you mean. The covolume is defined as the volume of any fundamental domain so it is necessary to provide a measure since the hypothesis we have on the ambient space `E` does not ensure that there is a canonical measure on it. It is true that, in most practical cases, `E` will be a `MeasureSpace` so we just want to use the volume of `E`. The tactic `volume_tac` does that automatically is `μ` is not provided.
> > 
> > 
> > So in practice if I have a lattice in, say, `ℂ` and I want to use the standard Lebesgue measure I have to write it explicitly?
> 
> Well, we could write `covolume L volume` (in fact, `MeasureTheory.MeasureSpace.volume`) or just `covolume L` (as it is written in `covolume_eq_det`) since Lean can infer automatically what the measure is since it is a `MeasureSpace`. It is only if you want to use an "unusual" measure (or the space `E` has no "canonical" measure) that would actually need to explicit provide it.

If `covolume L` works I am totally happy with it.

**Comment by riccardobrasca** [comment]
Thanks!

bors merge

**Pull request closed without merging**